strutil/shlex/shlex.go

```diff
@@ -414,3 +414,30 @@ func Split(s string) ([]string, error) {
 		subStrings = append(subStrings, word)
 	}
 }
+
+// Join returns a shell-escaped string version of the arguments in s.
+func Join(s []string) string {
+	quoted := make([]string, len(s))
+	for i, x := range s {
+		quoted[i] = Quote(x)
+	}
+	return strings.Join(quoted, " ")
+}
+
+// Quote returns a shell-escaped version of the single argument s.
+func Quote(s string) string {
+	if s == "" {
+		return `''`
+	}
+	for _, c := range s {
```

Comment:
This works fine, but also decodes the UTF-8 input into runes, and there's no reason to pay that penalty here. I suggest processing it as bytes instead. I think this would do it:

```
for i := 0; i < len(s); i++ {
    c := s[i]
    switch {
		case c >= 'a' && c <= 'z':
		case c >= 'A' && c <= 'Z':
		case c >= '0' && c <= '9':
		case strings.IndexByte("_@%+=:,./-", c) >= 0:
		...
    }
}

Reply:
Thanks for the insight! Update in https://github.com/canonical/x-go/pull/19/commits/08c9987a9a6f53281d401846358666dcb4c07a12.

Reply:
Looks okay with Ben's comments, but why are those not part of the PR still?

Reply:
Sorry about that @niemeyer.
I updated them in https://github.com/canonical/x-go/commit/08c9987a9a6f53281d401846358666dcb4c07a12, but I must have messed something up while rebasing. I should have checked carefully afterwards. I rebased and added the changes in https://github.com/canonical/x-go/pull/19/commits/58568dfafa951c44b6d1b05a01e9d2ab93e0c853.